pkg/upgrade/upgrade.go

```diff
@@ -819,501 +798,190 @@ func MigrateContainerSizesToHardwareProfiles(ctx context.Context, cli client.Cli
+					"opendatahub.io/display-name":                 "custom-serving",
+					"opendatahub.io/description":                  "",
+					"opendatahub.io/disabled":                     "false",
+					"opendatahub.io/managed":                      "false",
+				},
+			},
+			Spec: infrav1.HardwareProfileSpec{
+				Identifiers: []infrav1.HardwareIdentifier{
+					{
+						Identifier:   "cpu",
+						DisplayName:  "cpu",
+						ResourceType: "CPU",
+						MinCount:     intstr.FromInt(1),
+						DefaultCount: intstr.FromInt(1),
+					},
+					{
+						Identifier:   "memory",
+						DisplayName:  "memory",
+						ResourceType: "Memory",
+						MinCount:     intstr.FromString("1Gi"),
+						DefaultCount: intstr.FromString("1Gi"),
+					},
+				},
+			},
+		}
 
-	if effect, found, err := unstructured.NestedString(tolerationSettings, "effect"); err == nil && found {
-		toleration.Effect = corev1.TaintEffect(effect)
+		if err := cli.Create(ctx, hwp); err != nil {
+			if !k8serr.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create custom-serving HardwareProfile v1: %w", err)
+			}
+		}
+		log.Info("Successfully created custom-serving HardwareProfile", "namespace", namespace)
 	}
-
-	return []corev1.Toleration{toleration}, nil
+	return nil
 }
 
-func generateHardwareProfileFromAcceleratorProfile(ctx context.Context, ap unstructured.Unstructured, profileType string,
-	containerCounts map[string]string, notebooksOnlyToleration []corev1.Toleration) (*infrav1.HardwareProfile, error) {
+// AttachHardwareProfileToInferenceServices migrates AcceleratorProfile annotations from ServingRuntimes
+// and matches container sizes on InferenceServices to HardwareProfile annotations as described in RHOAIENG-33158.
+func AttachHardwareProfileToInferenceServices(ctx context.Context, cli client.Client, namespace string, odhConfig *unstructured.Unstructured) error {
 	log := logf.FromContext(ctx)
+	var multiErr *multierror.Error
 
-	// Extract AP fields
-	apName := ap.GetName()
-	apNamespace := ap.GetNamespace()
-
-	spec, found, err := unstructured.NestedMap(ap.Object, "spec")
-	if err != nil || !found {
-		return nil, errors.New("failed to get spec from AcceleratorProfile")
-	}
-
-	identifier, _ := spec["identifier"].(string)
-	displayName, _ := spec["displayName"].(string)
-	description, _ := spec["description"].(string)
-	enabled, _ := spec["enabled"].(bool)
-
-	// Create HWP name
-	hwpName := fmt.Sprintf("%s-%s", apName, profileType)
-
-	// Create annotations
-	annotations := map[string]string{
-		"opendatahub.io/dashboard-feature-visibility": GetFeatureVisibility(profileType),
-		"opendatahub.io/modified-date":                time.Now().Format(time.RFC3339),
-		"opendatahub.io/display-name":                 displayName,
-		"opendatahub.io/description":                  description,
-		"opendatahub.io/disabled":                     strconv.FormatBool(!enabled),
+	inferenceServices, err := getInferenceServices(ctx, cli, namespace)
```

Comment:
```suggestion
	inferenceServices, err := getInferenceServices(ctx, cli)
```
Need to list all inferenceservices cluster wide